### ISSUE TITLE
Support VRCFT OSC for Native Eye and Face tracking managers. 

### ISSUE THREAD

**Issue by sveken** [Yellow-Dog-Man/Resonite-Issues#1843]
### Is your feature request related to a problem? Please describe.

Currently as new headsets come out either community support for adding them via mods is required or time must be taken from the team to implement them Natively into Resonite.

### Describe the solution you'd like

Now that Resonite has the capability to handle OSC. Adding the option to support VRCFTs unified compressions parameters for direct face and eye tracking would open support up for pretty much all headsets and would leverage a much wider community for device support. 

Some manufactures like Pimax for example directly release modules for VRCFT  as their official social intergration for the Crystal. This would allow to piggy back off that work for eye tracking for that headset that is already done. 

### Describe alternatives you've considered

Currently the only method is to use the Mod VRCReciever however it does not drive the eye manager directly and relies on value copys.

### Additional Context

Related to https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1841 But instead using the data to directly drive the face and eye managers making it behave like a native supported headset.

**Comment by Geenz** [comment]
Pretty much everything in #975 applies as prerequisites to this, and would (to some extent) directly enable it as well for people who choose to use VRCFT.

Another (preferred) approach is OpenXR support more generally - which we have an internal version of ALXR to help bridge this gap until Sauce ships.

**Comment by Geenz** [comment]
For unified expressions - that would be a different set of work to enable that.  Presently we map everything to SRanipal’s blendshape parameters.

**Comment by sveken** [comment]
Yup unified expressions is just another naming standard for blendshapes for face/eye, reading into it more, it seems to be what vrcft maps all the different devices too to make the end result universal. However i belive it can be mapped to what is already in resonite.

As for what VRCFT can support eye/tracking wise, my understanding is it supports everything the device module supports as in the limitation is the device you are using. VRCFT is modular and every device has its own module that loads into it and provides tracking data, Pimax for example wrote their own official support module for it (for eye tracking), and there is a community module for the Crystal to support full range eye openness and dilation. Other projects like Project Babble (open source face tracking) also have modules.

The current official hardware support list is here https://docs.vrcft.io/docs/hardware/interface-compatibilities however there are alot of community modules for example even a steamlink one. 


With how VRCFT Maps to existing systems, the process is explained alot better on hazres mod page here https://github.com/hazre/VRCFTReceiver?tab=readme-ov-file#how-it-works but ill try explain here too. 

VRCFT normally reads  a json file that has all of the avatars set up parameters and what it wants to be tracked. This json file is copied to a location in appdata and on avatar load, tells vrcft to load that specific avatar .json via OSC. This file can be manually copied or in the mods case is auto generated on install. For Resonite the json file could contain all off the eye/face tracking data Resonite supports, an example is here https://github.com/hazre/VRCFTReceiver/blob/main/static/vrc_parameters.json VRCFTs doc page on OSC params are also here https://docs.vrcft.io/docs/tutorial-avatars/tutorial-avatars-extras/parameters.

Following the project it intends to be a universal interface for all new hardware whether by company or DIY. Adding compatibility would open up support for alot of current and future hardware.

**Comment by Geenz** [comment]
As noted - OSC would be needed first before we can support VRCFT.  Unified expressions is still its own separate work - notably the expression driver would need some work to support multiple standards for this sort of thing given the space is somewhat fragmented presently (Quest, SRanipal, VRCFT, Pico, Babble, etc.).  I’ve looked into it - and it’s not the most straightforward pile of work.

**Comment by FlameSoulis** [comment]
Couldn't some of it be simplified with plug-in libraries? Like, a simple, modular system that either the community could write as a step to allow internal processing of the facial systems or be officially composed? I get writing such a foundation would be adding some work, but given that the diversity of the standards doesn't seem to be getting slimmer (bear in mind, with the latest announcement of Horizon OS being made more open for other manufacturers, who knows what features will be put into the headsets and what standards they will follow).

**Comment by ko-tengu** [comment]
With OSC support added and VRCFT adding OSCQuery support, I do think this should end up with a higher priority. Directly interfacing with the wide range of hardware isn't a realistic goal for such a small team, and face tracking support using VRCFT as the only intended option is becoming more common with headsets that use these social features. There's a lot of people currently having to use mods because their face tracking doesn't align with either of the two options Resonite offers.

**Comment by Frooxius** [comment]
There are other approaches for this though. One of the things we'd like to do is actually modularize and open the device driver system, to allow for community contributions.

This would make it much easier for anyone to add support for more hardware, while fitting into Resonite's input system. This could include VRCFT driver as well.

We can add one natively, but I do not want the option to be limited only to that.

**Comment by ko-tengu** [comment]
I didn't say that VRCFT needs to be the only option, but having support would get you more time and flexibility to later add other options and have a fallback method that still works. 

The other problem with having to support a bunch of things is that it can cause a lot more fixing later down the line. We're seeing this happen already with the Steamlink-Quest Pro support where it's requiring some additional changes and it's breaking other people's attempts to make their own device drivers with mods.

**Comment by Frooxius** [comment]
What I'm saying is there are other approaches too and we might prefer to take those first, because they're more encompassing and flexible, rather than relying on a single set designed for another place.

What I'm proposing also fixes more of the issues you're mentioning. 

Consider it this way:
- If we implement VRCFT, then we support *only* VRCFT and we have to maintain it too - it too can break and cause issues
- If we implement the input system modularization, this allows for:
-- Community implemented VRCFT drivers if we don't make those officially before then
-- Community implemented drivers for other hardware
-- Community fixes for the drivers in case issues come up or modifications to customize them for everyone's needs

I understand your argument, but also understand that I really dislike relying on a single "standard" that then has to go into another standard. This goes against my principles on how I design and approach things with Resonite. Relying just on VRCFT makes me uncomfortable.

I see what I'm proposing as a more holistic and more encompassing approach to the problem.

**Comment by sveken** [comment]
Why would implementing vrcft force Resonite to only support vrcft? Would it be possible to not have it behind an enable checkbox? 
Some headsets come out of the box with vrcft support from the manufacture with the other other option being openXR.

I agree the modular input system does sound ideal. Though Resonite can already in its current state receive and send OSC data, just linking it to the eye/face drivers like steamlink. 

Speaking of steamlink, isn't it, itself just a middle man converting Metas direct OpenXR face/eye info into OSC data?

**Comment by Frooxius** [comment]
@sveken Because if we spend time implementing VRCFT support, then we only support VRCFT for handling lots of devices, because that's what we implemented.

If we spend time implementing the modularization, then we end up supporting more overall as result of those efforts.

---------------

The OSC receiving part doesn't really help much here - we just use library for that. That was never a complicated part of this process.

The main part is actually mapping those blendshapes to our system and adding mechanisms to properly detect and configure them. That's where bulk of the work goes.

**Comment by Frooxius** [comment]
To clarify, I'm not saying that we won't implement this ever. I'm just saying that to solve those particular problems you have mentioned, we might prioritize a different approach.

The VRCFT is being pushed as a solution for supporting lots of hardware and to avoid issues with individual drivers breaking here.

I would prefer us to address those specific problems by what I specified above, rather than relying in VRCFT to solve those, because that's more aligned with our approach and goals.

**Comment by Dusty-Sprinkles** [comment]
Maybe a better approach would just to be thinking of supporting VRCFT as a solution between solutions? for each solution that doesn't have its own custom thing yet- VRCFT, and work on adding things like MeowFace to fill the gaps, because theres a lot more niche solutions than would be reasonable compared to the time of just getting VRCFT imo

**Comment by Frooxius** [comment]
I don't see how that's better - that's exactly what I pointed out above that I'm uncomfortable with as an approach as relying on VRCFT for that.

If we are talking about expanding the system to allow generalized support, I'd rather invest time into opening and modularization of the system first - because that will allow for VRCFT to be added too with this, along with other drivers.

**Comment by ko-tengu** [comment]
Part of it for me is the somewhat less fair but still pretty understandable position that hardware that I own and can use isn't getting supported and the addition of this stuff is pretty slow (for understandable reasons). We get another kinda niche option added when wider used hardware is still unsupported and there's a solution for it in front of us.

Focusing an expandable driver system is also a perfectly acceptable path towards this, so long as it's something the community will be able to actually contribute and help with.

**Comment by sveken** [comment]
On the subject of mapping the names 
is more and more newer avatars have the unified expression naming  already as their blend shapes naming scheme so supporting that may have more benefits for newer avatars down the road as it becomes the default for new avatars.

**Comment by Frooxius** [comment]
But the solution I propose solves that too, in a way that I believe is healthier for Resonite.

We make it easy to load more drivers and we open source the existing drivers, to serve both as reference and to allow community customized versions.

**Comment by ProbablePrime** [comment]
Adding the blendshape names is over in: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2292 and should be treated differently than adding the driver for VRCFT.

#2292 , is just adding more blendshape names to our heuristics.

See also #2073

**Comment by FlameSoulis** [comment]
Just throwing my five cents into the ring, I have to agree that I'd rather have a modular API solution made that allows the community and the development team to write their own systems and allows more direct and specialized mapping options for certain conditions. This could also add a higher level of debugging for situations where an update to a service causes things to break, allowing the module to fail and better alert the logs where the failure occurred, while also allowing broken components to not hold the main project back should they no longer have any interest.
Speaking from maintaining the 3D Space Mouse module from the past, I'd rather have the option to add niche items myself or at someone's request than wishlisting the dev team a giant list of other things. This could also open the door for 'approved' modules.

**Comment by alexderpyfox** [comment]
Is this still planned?

**Comment by alexderpyfox** [comment]
When will vrcft be added

**Comment by dalekcoffee** [comment]
While I am a steamlink user myself, I agree this would help broader headset support and may even make the steam frame eye tracking with babble combo (that I think will be the likely future combo?) get some easier support maybe?
